src/kaleidoscope/plugin/OneShot.cpp

```diff
@@ -22,266 +22,339 @@
-  }
+// This function also isn't called anywhere by any core
+// plugins. Perhaps it should be deprecated.
+bool OneShot::isModifierActive(Key key) {
+  if (key < Key_LeftControl || key > Key_RightGui)
+    return false;
 
-  if (isOneShotKey_(mapped_key)) {
-    if (state_[idx].sticky) {
-      if (keyToggledOn(keyState)) {  // maybe on _off instead?
-        prev_key_ = mapped_key;
-        state_[idx].sticky = false;
-        cancelOneShot(idx);
-        should_cancel_ = false;
-      }
-    } else {
-      if (keyToggledOff(keyState)) {
-        state_[idx].pressed = false;
-        if (Runtime.hasTimeExpired(start_time_, hold_time_out)) {
-          cancelOneShot(idx);
-          should_cancel_ = false;
-        }
-      }
+  uint8_t n = key.getKeyCode() - Key_LeftControl.getKeyCode();
+  return state_[n].active;
+}
 
-      if (keyToggledOn(keyState)) {
-        state_[idx].pressed = true;
 
-        if (prev_key_ == mapped_key && isStickable(mapped_key)) {
-          uint16_t dtto = (double_tap_time_out == -1) ? time_out : double_tap_time_out;
-          if (!Runtime.hasTimeExpired(start_time_, dtto)) {
-            state_[idx].sticky = true;
-            prev_key_ = mapped_key;
-          }
-        } else {
-          start_time_ = Runtime.millisAtCycleStart();
+// ----------------------------------------------------------------------------
+// Plugin hook functions
 
-          state_[idx].position = key_addr.toInt();
-          state_[idx].active = true;
-          prev_key_ = mapped_key;
+EventHandlerResult OneShot::onKeyswitchEvent(
+  Key &mapped_key, KeyAddr key_addr, uint8_t key_state) {
 
-          activateOneShot(idx);
-        }
+  // Ignore injected key events. This prevents re-processing events
+  // that the hook functions generate (by calling `injectNormalKey()`
+  // via one of the `*OneShot()` functions). There are more robust
+  // ways to do this, but since OneShot is intended to react to only
+  // physical keypresses, this is adequate.
+  if (key_state & INJECTED)
+    return EventHandlerResult::OK;
+
+  // If it's not a OneShot key, and not a modifier, set up the active,
+  // non-sticky OneShot keys to be released at the end of the next
+  // cycle.
+  if (!isOneShotKey(mapped_key)) {
+    if (keyToggledOn(key_state)) {
+      prev_key_addr_ = key_addr;
+      if (!(mapped_key >= Key_LeftControl && mapped_key <= Key_RightGui) &&
+          !(mapped_key.getFlags() == (KEY_FLAGS | SYNTHETIC | SWITCH_TO_KEYMAP))) {
+        // We can't immediately send the release event(s) for active
+        // OneShot keys here, because that would release those keys
+        // before the report in which the keycodes they are meant to
+        // modify gets sent. Instead, we use this variable, which will
+        // get bit-shifted each cycle, and trigger the release when it
+        // gets to 1 (which is why it's written as a bit-shifted 1 here).
+        release_countdown_ = 1 << 1;
```

Comment:
I'm writing the value `2` here as `1 << 1` because it's going to be bit-shifted down once per cycle. When only the low bit is set, it will trigger the release of OneShot keys. Most of the time, it will be zero.